### PR TITLE
fix safe merge function

### PR DESCRIPTION
**Pull request by sodiumjoe**
## PR Checklist

- [x] Unit tests written (`common` at minimum)
- [x] PR has one of the `semver-` labels
- [x] Two core team engineer approvals

### CODE REVIEW

File: src/util/state-management.js [hunk 1]
```diff
@@ -70,16 +70,21 @@ export function getStatefulPropsContext(reducers, { getState, setState }) {
...
 		return srcValue;
 	}
 
+	// guards against traversing react elements which can cause cyclical recursion
 	// If we don't have this clause, lodash (as of 4.7.0) will attempt to
 	// deeply clone the react children, which is really freaking slow.
-	if (_.isArray(srcValue) && _.isUndefined(objValue)) {
+	if (isValidElement(srcValue)
+			|| (_.isArray(srcValue) && _.some(srcValue, isValidElement))
+			|| (_.isArray(srcValue) && _.isUndefined(objValue)) {
```
Comment (jondlm) [inline src/util/state-management.js]: travis barfed on this

**Comment by jondlm** [comment]
Can we also get some tests around this?

**Comment by sodiumjoe** [comment]
@jondlm yeah that's why i haven't checked the box yet

**Comment by sodiumjoe** [comment]
@jondlm k added tests

**Pull request merged**